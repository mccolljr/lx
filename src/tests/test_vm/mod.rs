use insta::{
    assert_snapshot,
    glob,
};

use std::io::Read;
use std::path::Path;
use std::process::{
    Command,
    Stdio,
};
use std::sync::Once;

static INIT_BUILD: Once = Once::new();

fn initialize() {
    INIT_BUILD.call_once(|| {
        assert!(Command::new("cargo")
            .arg("build")
            .stdout(Stdio::null())
            .stderr(Stdio::null())
            .status()
            .unwrap()
            .success());
    });
}

fn exec(path: &Path) -> Result<String, String> {
    let mut child = Command::new("./target/debug/lx")
        .arg(path)
        .stdout(Stdio::piped())
        .spawn()
        .expect("failed to spawn test");
    let exit_status = child.wait().expect("failed to run test");
    let mut output_str = String::new();
    child
        .stdout
        .expect("no stdout for test")
        .read_to_string(&mut output_str)
        .expect("failed to read stdout for test");
    if exit_status.success() {
        Ok(output_str)
    } else {
        Err(output_str)
    }
}

#[test]
fn test_scripts_success() {
    initialize();
    glob!("scripts_success/*.lx", |path| {
        println!("testing {:?}", path);
        match exec(path) {
            Ok(output) => {
                assert_snapshot!(output);
            }
            Err(output) => {
                panic!("unexpected error:\n{}", output);
            }
        }
    });
}

#[test]
fn test_scripts_error() {
    initialize();
    glob!("scripts_error/*.lx", |path| {
        println!("testing {:?}", path);
        match exec(path) {
            Ok(output) => {
                panic!("expected failure, but got success:\n{}", output);
            }
            Err(output) => {
                assert_snapshot!(output);
            }
        }
    });
}
