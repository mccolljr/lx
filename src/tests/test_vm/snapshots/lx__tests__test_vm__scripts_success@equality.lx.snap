---
source: src/tests/test_vm/mod.rs
expression: output
input_file: src/tests/test_vm/scripts_success/equality.lx
---
1 == 1.0:             true
1.0 == 1:             true
1.1 == 1:             false
1 == 1.1:             false
'a' == 'a':           true
'a' == 'b':           false
'b' == 'a':           false
'b' + 'a' == 'ba':    true
true == true:         true
false == false:       true
test == test:         true
test == alias_test:   true
alias_test == test:   true
test == test2:        false
test2 == test:        false
obj1 == obj1:         true
obj1 == obj2:         true
obj2 == obj1:         true
arr1 == arr1:         true
arr1 == arr2:         true
arr2 == arr1:         true
print == print:       true
print == alias_print: true
OK!

